ddsrouter_event/test/unittest/signal/SignalHandlerTest.cpp

```diff
@@ -0,0 +1,327 @@
+            {
+                calls++;
+            } );
+
+    // Raise signal
+    raise(SIGNAL_SIGINT);
+
+    // Force handler to wait for signal
+    handler.wait_for_event();
+
+    // Check that signal has been received
+    ASSERT_EQ(1, calls);
+}
+
+/**
+ * Receive a signal from a Signal handler with complex environment
+ *
+ * CASES:
+ * - SIGINT double handler
+ * - SIGINT double signal
+ * - SIGINT receive signal after unset
+ * - SIGINT receive signal just before destroying
+ */
+TEST(SignalHandlerTest, receive_signal)
+{
+    // SIGINT double handler
+    {
+        // Number of calls to the signal
+        uint32_t calls = 0;
+
+        // Create signal handler
+        SignalHandler<SIGNAL_SIGINT> handler1 ( [&calls](int /* signal_number */ )
+                {
+                    calls++;
+                } );
+        SignalHandler<SIGNAL_SIGINT> handler2 ( [&calls](int /* signal_number */ )
+                {
+                    calls++;
+                } );
+
+        // Raise signal
+        raise(SIGNAL_SIGINT);
+
+        // Force handler to wait for signal
+        handler1.wait_for_event();
+        handler2.wait_for_event();
+
+        // Check that signal has been received
+        ASSERT_EQ(2, calls);
+    }
+
+    // SIGINT double signal
+    {
+        // Number of calls to the signal
+        uint32_t calls = 0;
+
+        // Create signal handler
+        SignalHandler<SIGNAL_SIGINT> handler( [&calls](int /* signal_number */ )
+                {
+                    calls++;
+                } );
+
+        // Raise signal
+        raise(SIGNAL_SIGINT);
+        raise(SIGNAL_SIGINT);
+
+        // Force handler to wait for signal
+        handler.wait_for_event(2);
+
+        // Check that signal has been received
+        ASSERT_EQ(2, calls);
+    }
+
+    // SIGINT receive signal after unset
+    {
+        // Number of calls to the signal
+        uint32_t calls = 0;
+
+        // Create signal handler
+        SignalHandler<SIGNAL_SIGINT> handler( [&calls](int /* signal_number */ )
+                {
+                    calls++;
+                } );
+
+        // Raise signal
+        raise(SIGNAL_SIGINT);
+
+        // Force handler to wait for signal
+        handler.wait_for_event();
+
+        // Check that signal has been received
+        ASSERT_EQ(1, calls);
+
+        // Unset callback, so new signal raise is not listened
+
+        handler.unset_callback();
+
+        // Raise signal
+        raise(SIGNAL_SIGINT);
+
+        // Check that signal has not been received
+        ASSERT_EQ(1, calls);
+    }
+
+    // SIGINT receive signal just before destroying
+    {
+        {
+            // Create signal handler
+            SignalHandler<SIGNAL_SIGINT> handler( [](int /* signal_number */ )
+                    {
+                        /* empty callback */ } );
+        }
+
+        // Destroying handler while Raise signal
+        raise(SIGNAL_SIGINT);
+    }
```

Comment:
Maybe it could be tested the reception of a signal while destroying signal handler/unsetting callback, but I don't know if that's possible.

Perhaps create a Test class overriding the destructor/unset_callback_nts and introduce a `raise(SIGNAL_SIGINT)` somewhere in.

Reply:
this test was implemented wrong, check it now. It is a race condition so cannot be forced.

The way you suggest changes radically the behaviour of the class, and we do not have certainty of what `raise` does, so it can break the test in other ways